autopilot/choice_test.go

```diff
@@ -13,6 +13,101 @@ var (
 	nID4 = NodeID([33]byte{4})
 )
 
+// TestWeightedChoiceEmptyMap tests that passing in an empty slice of weights
+// returns an error.
+func TestWeightedChoiceEmptyMap(t *testing.T) {
+	t.Parallel()
+
+	var w []float64
+	_, err := weightedChoice(w)
+	if err != ErrNoPositive {
+		t.Fatalf("expected ErrNoPositive when choosing in "+
+			"empty map, instead got %v", err)
+	}
+}
+
+// TestWeightedChoiceSingleIndex tests that choosing randomly in a slice with one
+// positive element always returns that one index.
+func TestWeightedChoiceSingleIndex(t *testing.T) {
+	t.Parallel()
+
+	w := []float64{0, 0.5, 0, 0}
+	for i := 0; i < 1000; i++ {
```

Comment:
We can modify this slightly to use `testing.Quick` instead. It would look something like: 
```go
property := func(weights [10]float64) bool {
    allButOneZero := func() (bool, float64) {
        var (
           numZero uint32
           nonZero float64
        )
       
        for _, w := range weights {
          if w != 0 {
             numZero++
             nonZero = w
         }
       }
       
      return numZero == 1, nonZero
    }
  
   conditionMet, nonZeroElem := allButOneZero()
   if conditionMet {
    choice, err := weightedChoice(w) 
    if err != nil {
      return false
    }
    return choice == nonZeroEleme
   }
}

if err := quick.Check(property, nil); err != nil {
    t.Fatalf(err)
}
```

Reply:
The above example illustrates how to write property based tests in Go using `testing/quick`. Given a set of random (or constrained data via the config), we make a predicate function that targets a particular property. In this case, the property is: if there's a single non-zero element, then we should always select that. I made it a fixed slice to make the code a bit shorter, but we can also accept a `[]float64` and use the config/generator to restrict the form of the input data.

Reply:
Done!